Hackage revisions don't seem to be respected

I use [arbtt](http://arbtt.nomeata.de/#what) and since it was marked broken in nixpkgs I packaged it using haskell.nix:

```
==> ~/src/dotfiles/opt/arbtt/default.nix <==
{ pkgs ? import ~/src/nix-config/haskell.nix/nixpkgs.nix
, compiler-nix-name ? "ghc8102"
}:
(pkgs.haskell-nix.hackage-package {
  inherit compiler-nix-name;
  name = "arbtt";
  version = "0.10.2";
  index-state = "2020-10-26T22:14:30Z";
  configureArgs = "--constraint 'strict <0.4'";
}).components.exes

==> ~/src/nix-config/haskell.nix/nixpkgs.nix <==
let
  haskellNix = import (builtins.fetchTarball
    ( "https://github.com/input-output-hk/haskell.nix/archive/"
    + "5f80ca910b0c34562f76aa4dcfc2464840b2d0ef.tar.gz"
    )) {};
in
  haskellNix.pkgs
```

I ended up having to include the `configureArgs` because arbtt doesn't build with "strict" version 0.4, it results in precedence parsing errors:

```
[ 8 of 15] Compiling LeftFold         ( src/LeftFold.hs, dist/build/arbtt-capture/arbtt-capture-tmp/LeftFold.o )

src/LeftFold.hs:80:70: error:
    Precedence parsing error
        cannot mix ‘:!:’ [infix 2] and ‘:!:’ [infix 2] in the same infix expression
   |
80 | runOnGroups eq (LeftFold sti pi fi) (LeftFold sto po fo) = LeftFold (S.Nothing :!: sti :!: sto) go finish 
   |                                                                      ^^^^^^^^^^^^^^^^^^^^^^^^^
```

However looking into this further I found out it's a [known and fixed issue issue upstream](https://bitbucket.org/nomeata/arbtt/issues/64/cabal-install-arbtt-not-working) and [fixed on hackage since early september with a revision](https://hackage.haskell.org/package/arbtt-0.10.2/revisions/).

This only started happening since I bumped my haskell.nix revision (and GHC version but that doesn't seem to make a difference), as I'd successfully built this earlier this month so this might be a regression.

--- a/overlays/haskell.nix
+++ b/overlays/haskell.nix
@@ -471,7 +471,8 @@ final: prev: {
                   url = "https://hackage.haskell.org/package/${name}-${version}/revision/${toString rev.revNum}.cabal";
                   inherit (rev) sha256;
                 };
-            in let src = final.buildPackages.pkgs.runCommand "${name}-${version}-src" { } (''
+                revSuffix = final.lib.optionalString (rev.revNum > 0) "-r${toString rev.revNum}";
+            in let src = final.buildPackages.pkgs.runCommand "${name}-${version}${revSuffix}-src" { } (''
                   tmp=$(mktemp -d)
                   cd $tmp
                   tar xzf ${tarball}

--- a/overlays/haskell.nix
+++ b/overlays/haskell.nix
@@ -460,20 +460,28 @@ final: prev: {
             { name
             , compiler-nix-name
             , version
+            , revision ? "default"
             , ... }@args':
             let args = { caller = "hackage-project"; } // args';
                 tarball = final.pkgs.fetchurl {
-                url = "mirror://hackage/${name}-${version}.tar.gz";
-                inherit (hackage.${name}.${version}) sha256; };
-            in let src = final.buildPackages.pkgs.runCommand "${name}-${version}-src" { } ''
-                tmp=$(mktemp -d)
-                cd $tmp
-                tar xzf ${tarball}
-                mv "${name}-${version}" $out
-                '';
+                  url = "mirror://hackage/${name}-${version}.tar.gz";
+                  inherit (hackage.${name}.${version}) sha256; };
+                rev = hackage.${name}.${version}.revisions.${revision};
+                cabalFile = final.pkgs.fetchurl {
+                  url = "https://hackage.haskell.org/package/${name}-${version}/revision/${toString rev.revNum}.cabal";
+                  inherit (rev) sha256;
+                };
+            in let src = final.buildPackages.pkgs.runCommand "${name}-${version}-src" { } (''
+                  tmp=$(mktemp -d)
+                  cd $tmp
+                  tar xzf ${tarball}
+                  mv "${name}-${version}" $out
+                '' + final.lib.optionalString (rev.revNum > 0) ''
+                  cp ${cabalFile} $out/${name}.cabal
+                '');
           in cabalProject' (
             (final.haskell-nix.hackageQuirks { inherit name version; }) //
-              builtins.removeAttrs args [ "version" ] // { inherit src; });
+              builtins.removeAttrs args [ "version" "revision" ] // { inherit src; });
 
         # This function is like `cabalProject` but it makes the plan-nix available
         # separately from the hsPkgs.  The advantage is that the you can get the

